File: Graph/ts/hasCycle.ts
import { Graph } from "./types";

export function hasCycle(graph: Graph): boolean {
  const visiting = new Set<string>();
  const visited = new Set<string>();

  for (const node of Object.keys(graph)) {
    if (detect(graph, node, visiting, visited)) {
      return true;
    }
  }
  return false;
}

function detect(graph: Graph, node: string, visiting: Set<string>, visited: Set<string>): boolean {
  if (visited.has(node)) return false;
  if (visiting.has(node)) return true;

  visiting.add(node);
  for (const neighbor of graph[node] ?? []) {
    if (detect(graph, neighbor, visiting, visited)) {
      return true;
    }
  }

  visiting.delete(node);
  visited.add(node);
  return false;
}

File: Graph/csharp/HasCycle.cs
using System.Collections.Generic;

namespace GraphSolutions;

public static class HasCycle
{
    public static bool Solve(Dictionary<string, List<string>> graph)
    {
        var visiting = new HashSet<string>();
        var visited = new HashSet<string>();

        foreach (var node in graph.Keys)
        {
            if (Detect(graph, node, visiting, visited))
            {
                return true;
            }
        }

        return false;
    }

    private static bool Detect(Dictionary<string, List<string>> graph, string node, HashSet<string> visiting, HashSet<string> visited)
    {
        if (visited.Contains(node))
        {
            return false;
        }
        if (visiting.Contains(node))
        {
            return true;
        }

        visiting.Add(node);
        if (graph.TryGetValue(node, out var neighbors))
        {
            foreach (var neighbor in neighbors)
            {
                if (Detect(graph, neighbor, visiting, visited))
                {
                    return true;
                }
            }
        }

        visiting.Remove(node);
        visited.Add(node);
        return false;
    }
}
